gcm/gcm_connector_test.go

```diff
@@ -106,13 +195,238 @@ func TestRemoveTailingSlash(t *testing.T) {
+	assert.Nil(err)
+	assert.Equal(len(gcm.stopChan), 1, "StopChan")
+
+}
+
+func TestGcmConnector_StartWithMessageSending(t *testing.T) {
+	defer initCtrl(t)()
+
+	assert := assert.New(t)
+	routerMock := NewMockRouter(ctrl)
+	routerMock.EXPECT().Subscribe(gomock.Any()).Do(func(route *server.Route) {
+		assert.Equal("/gcm/broadcast", string(route.Path))
+		assert.Equal("gcm_connector", route.UserID)
+		assert.Equal("gcm_connector", route.ApplicationID)
+	})
+
+	kvStore := store.NewMemoryKVStore()
+	routerMock.EXPECT().KVStore().Return(kvStore, nil)
+
+	gcm, err := NewGCMConnector(routerMock, "/gcm/", "testApi")
+	assert.Nil(err)
+
+	err = gcm.Start()
+	assert.Nil(err)
+
+	done := make(chan bool, 1)
+	mockSender := createSender(RoundTripperFunc(func(req *http.Request) *http.Response {
+		// signal the ending of processing
+		defer func() {
+			done <- true
+		}()
+
+		return composeHTTPResponse(req, http.StatusOK, correctGcmResponseMessageJSON)
+	}))
+	gcm.sender = mockSender
+
+	// put a broadcast message with no recipients and expect to be dropped by
+	broadcastMsgWithNoRecipients := server.MsgAndRoute{Message: &protocol.Message{ID: uint64(4), Body: []byte("{id:id}"), Time: 1405544146, Path: "/gcm/broadcast"}}
+	gcm.channelFromRouter <- broadcastMsgWithNoRecipients
+	time.Sleep(time.Millisecond * 1000)
+	// expect that the HTTP Dummy Server to not handle any requests
+
+	// put a dummy gcm message with minimum information
+	msgWithNoRecipients := server.MsgAndRoute{Message: &protocol.Message{ID: uint64(4), Body: []byte("{id:id}"), Time: 1405544146, Path: "/gcm/marvin/gcm124/subscribe/stuff"}, Route: &server.Route{ApplicationID: "id"}}
+	gcm.channelFromRouter <- msgWithNoRecipients
+	// expect that the Http Server to give us a malformed message
+	<-done
+
+	//wait a couple of seconds and  Stop the GcmConnector
+	time.Sleep(time.Millisecond * 2000)
+	err = gcm.Stop()
+	assert.Nil(err)
+}
+
+func TestGCMConnector_BroadcastMessage(t *testing.T) {
+	defer initCtrl(t)()
+
+	a := assert.New(t)
+
+	// given:  a rest api with a message sink
+	routerMock := NewMockRouter(ctrl)
+
+	kvStore := store.NewMemoryKVStore()
+	routerMock.EXPECT().KVStore().Return(kvStore, nil)
+
+	routerMock.EXPECT().Subscribe(gomock.Any()).Do(func(route *server.Route) {
+		a.Equal("/notifications", string(route.Path))
+		a.Equal("marvin", route.UserID)
+		a.Equal("gcmId123", route.ApplicationID)
+	})
+
+	gcm, err := NewGCMConnector(routerMock, "/gcm/", "testApi")
+	a.Nil(err)
+
+	url, _ := url.Parse("http://localhost/gcm/marvin/gcmId123/subscribe/notifications")
+	// and a http context
+	req := &http.Request{URL: url, Method: "POST"}
+	w := httptest.NewRecorder()
+
+	// when: I POST a message
+	gcm.ServeHTTP(w, req)
+
+	// the the result
+	a.Equal("registered: /notifications\n", string(w.Body.Bytes()))
+
+	done := make(chan bool, 1)
+	mockSender := createSender(RoundTripperFunc(func(req *http.Request) *http.Response {
+		// signal the ending of processing
+		defer func() {
```

Comment:
I think this can be made even shorter, if you make `composeHTTPResponse` to accept a channel and return the final `RoundTripperFunc`